### ISSUE TITLE
Knowledge Base is dead on the frozen Windows build: the same off-main-thread native import deadlocks at embedding-model download

### ISSUE THREAD

**Issue by yolaatar** [erudi-app/erudi#321]
Sibling of #313, filed with the dump @rayanhanader asked for. His structural read was right, and the scope is bigger than "chunking.py might be slow".

## It deadlocks, and earlier than expected

Not at ingestion. At the **embedding model download**, before a single document is touched. Fresh install, clean env, Knowledge Base tab, click Download on the embedding gate. The UI sits on "Downloading the embedding model… This can take a minute" indefinitely (observed 7+ minutes).

```
Thread 28036 (idle): "e5-download"
    _call_with_frames_removed (<frozen importlib._bootstrap>:488)
    create_module (<frozen importlib._bootstrap_external>:1293)
    ...
    <module> (blas.py:247)              <- scipy BLAS native module
    <module> (_chunking.py:11)          <- sklearn
    ...
    _load_model (embedding_model.py:109)
    _run_download (embedding_model.py:153)
    run (threading.py:1012)
    _bootstrap_inner (threading.py:1075)
```

Zero forward progress, same signature as #313:

```
cpu_delta = 0.000s
mem_delta = 0.05 MB      (15s window)
```

## The important new fact: it is not an anyio/threadpool thing

#313 dies on an `AnyIO worker thread` via `run_in_threadpool`. This one dies on a **plain `threading.Thread`** (`e5-download`, started by `_run_download`). Same `LoadLibraryExW` on the same scipy/sklearn BLAS chain.

So the trigger is not FastAPI's threadpool or anything anyio does. It is **any non-main thread** loading these native modules under PyInstaller's frozen importer. That broadens the diagnosis from "get the probe off the request path" to "off-main-thread native imports are unsafe in this build, full stop".

Worth noting while reading dumps: `run.py` runs uvicorn inside `Thread-1 (run_server)` and leaves `MainThread` supervising at `run.py:481`. So essentially all application work already happens off the main thread, and any lazy first-import of this dependency chain is exposed, wherever it lives.

## What this does to the #313 plan

Direction 2 (GGUF KV + jinja2) removes `transformers` from the chat probe and fixes chat. It does **not** help here: the KB genuinely needs `sentence_transformers`, so there is nothing to remove. A build carrying only direction 2 ships working chat and a Knowledge Base that hangs on first use.

`OPENBLAS_NUM_THREADS=1` does not help either, tested on the #313 repro (see that thread).

## The one lead left, and the question that decides it

If the deadlock is specifically about loading these DLLs off the main thread, then importing them **on the main thread** should be safe, and that is a class-wide cure covering chat, KB, and anything added later.

The open question is what that costs. Known: the same import from a worker thread in a dev venv takes 5.8s. Unknown: how long a **main-thread** import takes in the frozen build, which is the number that decides whether a startup pre-warm is acceptable or whether it has to be lazy-but-main-thread-marshalled.

Note this would have to be a real main-thread import, not the post-ready background task pattern used by `backfill_wire_tools_startup`: that runs off the main thread and would deadlock exactly like `e5-download` does.

## Repro

1. Install the Windows build, no KB created yet
2. Knowledge Base tab, click Download on the embedding gate
3. It never completes. `py-spy dump --pid <backend>` shows the frame above

**Comment by rayanhanader** [comment]
The sourced root-cause investigation landed on #313 (https://github.com/djaxchi/erudi/issues/313#issuecomment-5347194386) and it covers this issue directly: the leading candidate is a duplicate-CRT split-brain created by the bundled `ucrtbase.dll`/`msvcp140.dll` in `_internal` (PyInstaller collects them by design; the MinGW-built scipy OpenBLAS binds to the bundled copy while the MSVC world uses the system one). If the 2-minute no-rebuild experiment described there (move the two unmangled CRT DLLs out of `_internal`, rerun this KB repro) makes the hang disappear, the fix is a spec-level dedup — class-wide, and `sentence_transformers` stays exactly where it is, no import relocation needed. Worth running the experiment against this repro too since it is the earliest trigger.

**Comment by yolaatar** [comment]
Measured the main-thread question with a standalone PyInstaller onedir probe (same `--onedir` shape as `backend.spec`), so no app rebuild. Timings for the same import, main thread vs a plain worker thread:

| target | main thread | worker thread |
|---|---|---|
| `scipy.linalg.blas` | 0.19s | 2.08s |
| `sklearn.utils._chunking` | 0.75s | 4.69s |
| `transformers.AutoTokenizer` | **4.50s** | 20.75s |
| `transformers` after `nvmlInit()` | not run | **40.22s** |

Unfrozen control in the dev venv: 0.28s on a worker thread. So the penalty is frozen-specific, as expected.

Two things worth taking from this.

**1. The minimal probe does NOT deadlock.** Even `transformers` on a worker thread with NVML preloaded completes, just slowly. So "frozen + worker thread + transformers" is not sufficient on its own; the hard block needs more of the real bundle's native state (torch CUDA, faiss, llama.cpp, psycopg/pgserver). What the probe does show is a clean monotonic trend: the more native DLLs already loaded, the worse the off-main-thread import gets. NVML alone doubles it. That is consistent with loader-lock contention that tips into a permanent block once the real app's full DLL set is resident, which matches the zero-CPU signature we measured in-app.

**2. Main thread never degraded, in any configuration.** It is 4-5x faster than the worker thread in every target, and it is the only mode that was immune throughout.

So the answer to "what does a main-thread import cost in the frozen build" is **~4.5s for transformers in a minimal bundle**. Treat that as a lower bound: the real bundle is much larger and will be slower. But it is seconds, not minutes, which is the thing that was unknown, and it makes a main-thread pre-warm at startup viable rather than the "several minutes of startup" worry from #313.

Design consequence: the pre-warm has to be a genuine main-thread import. `MainThread` is currently parked in `run.py:481` supervising while uvicorn runs in `Thread-1`, so there is a natural place to do this before the server thread starts, and it is the one context the probe never saw degrade. The post-ready background task pattern remains unusable for exactly the reason noted above.

Probe source and both binaries are reproducible in a few minutes if anyone wants to re-run against a bundle closer to the real one (adding torch+CUDA to the probe is the obvious next refinement, and would likely be what finally reproduces the hard deadlock outside the app).

**Comment by rayanhanader** [comment]
That probe is the right instrument, and it turns the unknown into a number. Three things fall out of your data, one of which flips the obvious design choice.

## 1. Your own curve says the pre-warm must be EARLY, not after ready

The intuitive placement is "after the port is bound, while `MainThread` is parked in the join loop" — it costs zero perceived startup time. Your measurements say that is the **worst** possible spot.

The trend you found is monotonic in resident native DLLs: NVML alone doubles the transformers import (20.75s -> 40.22s on a worker). By the time `ready` is emitted, the FastAPI lifespan has already run: engine selection, hardware detection (which is what calls `nvmlInit()`), torch, pgserver, the embedded Postgres cluster. That is precisely the maximally-loaded state, so a pre-warm there pays the top of your curve — and on the real bundle, which is much larger than the probe, the top of that curve is where the hard block lives.

The cheap spot is the opposite end: **the first thing in `main()`, before the app imports drag anything in**. That is your ~4.5s lower bound rather than your 40s figure, and it is trivially main-thread since nothing else has started.

## 2. The startup budget absorbs it

`run.py` allows 120s normally and 300s on first run (`STARTUP_TIMEOUT_SECONDS` / `FIRST_RUN_TIMEOUT_SECONDS`), and first boot already spends ~17s on the one-time initdb. So a 5s — even 30s — pre-warm before binding the port has plenty of headroom and cannot trip `PORT_TIMEOUT`. Worth emitting the `starting` event before it so the Electron splash is not staring at silence.

## 3. Heads-up: `run.py` moved an hour ago

PR #322 (#224) landed and touches exactly this area, so rebase before writing the pre-warm:

- `main()` now captures `initial_ppid = os.getppid()` on its **first line** — keep the pre-warm after that one line (it must stay ahead of the heavy imports for the reason in #224).
- There is now a parent-death watchdog on its own daemon thread. Good news for this design: it is independent of `MainThread`, so a blocking pre-warm cannot delay parent-death detection.
- But it does confirm the "after ready" placement is bad on a second axis: the join loop at `run.py:612-615` is what emits the `shutdown` event, so blocking `MainThread` there for tens of seconds would delay clean shutdown reporting by exactly that long.

## 4. Your probe can now settle the duplicate-CRT hypothesis cheaply

The most useful property of what you built is that it gives a **continuous** signal (import duration) instead of the binary one (deadlocks / does not). You no longer need to reproduce the hard deadlock to test a mechanism — you can test it against the 11x worker-thread penalty you already measured on `scipy.linalg.blas` (0.19s -> 2.08s).

So: run the same probe twice, once as built and once with the unmangled `ucrtbase.dll` and `msvcp140.dll` removed from its `_internal`. If the split-brain CRT is the mechanism, the worker-thread penalty should collapse toward the main-thread number. If it does not move, my hypothesis is dead and the pre-warm becomes the whole fix rather than a belt-and-braces.

Being fair to your data: the minimal probe not deadlocking does weaken "duplicate CRT alone is sufficient". It stays consistent with "contention across many native DLLs, tipping into a permanent block once two CRT instances are in play" — but that is exactly what the two-run comparison decides, on a bench you already have, in minutes. Adding torch+CUDA to the probe as you suggest is the other obvious refinement, and the two combine well: if torch+CUDA finally reproduces the hard deadlock, the CRT-removal run on that same bundle becomes a direct yes/no on the fix.

**Comment by yolaatar** [comment]
Root-caused and fixed in #337.

This is not a KB problem and needs no import surgery. The `e5-download` thread deadlocks because the **stdin-EOF watcher** (`run.py:299`) parks a daemon thread in a blocking `os.read` on fd 0 for the life of the process. On Windows that enters UCRT `_read()`, which takes the per-fd lock and then blocks in `ReadFile`; while it sits there, any off-main-thread native import of the OpenBLAS chain hangs in `LoadLibraryExW`.

Bisected on the packaged 2.0.0 CUDA build, same bundle throughout: four runs with `ERUDI_WATCH_STDIN` unset completed the download, three runs with it set deadlocked with identical frames and zero progress. Electron sets it at `main.js:261`; nothing else does. Table and dumps are in #337 and #313.

Two consequences for the plan here:

- **The main-thread pre-warm is not needed.** It was the right idea against the wrong model of the bug (a slow import that had to be paid somewhere). The import is not slow off the main thread, it is blocked, and once the watcher stops parking in the CRT it imports normally wherever it runs. That also retires the startup-cost tradeoff @rayanhanader and I were weighing on the 0.19s/2.08s numbers.
- **`sentence_transformers` stays exactly where it is**, lazily imported inside `_load_model`, which is the class-wide outcome we wanted.

Verification still needs the rebuild, since `run.py` is frozen into the current bundle. Keeping this open until the packaged build confirms it.

**Comment by yolaatar** [comment]
**Fixed, confirmed on a real packaged build.**

Built locally from `origin/main` + #337 only and run through the actual Electron app, so `ERUDI_WATCH_STDIN=1` is set as in production.

```
20:18:59Z  Downloading embedding model intfloat/multilingual-e5-small -> ...\models_cache
20:19:31Z  Embedding model download complete (32.0s)
```

Previously this thread parked in `create_module` -> `blas.py:247` and never returned (zero progress past 200s across repeated runs). The `sentence_transformers` import now completes normally on the same plain `threading.Thread`, with no import surgery, no pre-warm, and the module left exactly where it was.

Full evidence and the shutdown-contract regression check are on #337. Closing once that merges.
